packages/compass-query-bar/src/stores/atlas-signin-reducer.ts

```diff
@@ -0,0 +1,208 @@
+  type: AtlasSignInActions.RestoringStart;
+  id: number;
+};
+
+export type AtlasSignInRestoringFailedAction = {
+  type: AtlasSignInActions.RestoringFailed;
+};
+
+export type AtlasSignInRestoringSuccessAction = {
+  type: AtlasSignInActions.RestoringSuccess;
+};
+
+export type AtlasSignInStartAction = {
+  type: AtlasSignInActions.Start;
+  id: number;
+};
+
+export type AtlasSignInSuccessAction = {
+  type: AtlasSignInActions.Success;
+};
+
+export type AtlasSignInErrorAction = {
+  type: AtlasSignInActions.Error;
+  error: string;
+};
+
+export type AtlasSignInCancelAction = { type: AtlasSignInActions.Cancel };
+
+const INITIAL_STATE = {
+  currentAttemptId: -1,
+  state: 'initial' as const,
+  error: null,
+  user: null,
+};
+
+const AbortControllerMap = new Map<number, AbortController>();
+
+let abortId = 0;
+
+function getAbortSignal() {
+  const id = abortId++;
+  const controller = new AbortController();
+  AbortControllerMap.set(id, controller);
+  return { id, signal: controller.signal };
+}
+
+function abort(id: number) {
+  AbortControllerMap.get(id)?.abort();
+  AbortControllerMap.delete(id);
+}
+
+const reducer: Reducer<AtlasSignInState> = (
+  state = { ...INITIAL_STATE },
+  action
+) => {
+  if (
+    isAction<AtlasSignInRestoringStartAction>(
+      action,
+      AtlasSignInActions.RestoringStart
+    )
+  ) {
+    return { ...state, state: 'restoring', currentAttemptId: action.id };
+  }
+
+  if (
+    isAction<AtlasSignInRestoringSuccessAction>(
+      action,
+      AtlasSignInActions.RestoringSuccess
+    )
+  ) {
+    return { ...state, state: 'success', error: null };
+  }
+
+  if (
+    isAction<AtlasSignInRestoringFailedAction>(
+      action,
+      AtlasSignInActions.RestoringFailed
+    )
+  ) {
+    return { ...state, state: 'unauthenticated' };
+  }
+
+  if (isAction<AtlasSignInStartAction>(action, AtlasSignInActions.Start)) {
+    return { ...state, state: 'in-progress', currentAttemptId: action.id };
+  }
+
+  if (isAction<AtlasSignInSuccessAction>(action, AtlasSignInActions.Success)) {
+    return { ...state, state: 'success', error: null };
+  }
+
+  if (isAction<AtlasSignInErrorAction>(action, AtlasSignInActions.Error)) {
+    return { ...state, state: 'error', user: null, error: action.error };
+  }
+
+  if (isAction<AtlasSignInCancelAction>(action, AtlasSignInActions.Cancel)) {
+    return { ...INITIAL_STATE, state: 'canceled' };
+  }
+
+  return state;
+};
+
+export const getInitialSignInState = (): QueryBarThunkAction<Promise<void>> => {
+  return async (dispatch, getState, { atlasService }) => {
+    if (getState().atlasSignIn.state === 'success') {
+      return;
+    }
+    dispatch(cancelSignIn());
+    const { id, signal } = getAbortSignal();
+    dispatch({ type: AtlasSignInActions.RestoringStart, id });
+    dispatch({
+      type: (await atlasService.isAuthenticated({ signal }).catch(() => {
+        // For the initial state check if failed to check auth for any reason we
```

Comment:
This should check if we caught a cancellation error and not dispatch in that case